### PR TITLE
Manual scale limits default to present names of values

### PR DESCRIPTION
**Pull request by teunbrand**
#### Why this PR?

In ggplot2 v3.3.4, #4471 was introduced to allow fewer items in `values` than auto-detected breaks.
This has had some consequences, most notably that an extra legend key was produced if a named `values` was longer than auto-detected `breaks` (#4511, #4534, #4556, #4545). One workaround is currently to use `limits = force`, which feels unneccessary (https://github.com/tidyverse/ggplot2/issues/4511#issuecomment-866185530). Additionally, the ordering of the legend keys was effected by the names in the `values` argument (#4616).

#### What does this PR do?

This PR aims to revert back to the old behaviour (in most part), while preserving the spirit of #4471. It does so by replacing a vector-based `limits` by a function-based `limits` that intersects the names of the values by the observed limits.

#### Foreseen problems

Both this PR and #4547 aim to do the same thing and therefore these PRs are in conflict. This PR solely affects manual scales, whereas all scales are affected in #4547.

#### Examples

``` r
library(ggplot2)

p <- ggplot(mpg, aes(displ, hwy, colour = drv)) +
  geom_point(size = 2)

values <- c("f" = "dodgerblue", "r" = "tomato", "4" = "limegreen", 
            "dummy" = "black")

# Does not change default order (#4616), nor produces extra key (#4511)
p + scale_colour_manual(
    values = values
  )
```

``` r
# Can add extra key by setting limits
# The plot is the same as CRAN version without limits argument
p + scale_colour_manual(
  values = values,
  limits = names(values)
)
```

``` r
# Can set undefined values to na.value (spirit of #4471)
p + scale_colour_manual(
  values = values[1:2]
)
```

``` r
# Can set fewer keys by setting breaks (same as CRAN version)
p + scale_colour_manual(
  values = values,
  breaks = names(values)[1:2]
)
```

<sup>Created on 2021-09-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### CODE REVIEW

File: R/scale-manual.r [hunk 1]
```diff
@@ -132,8 +132,9 @@ manual_scale <- function(aesthetic, values = NULL, breaks = waiver(), ..., limit
     force(values)
   }
 
-  if (is.null(limits)) {
-    limits <- names(values)
+  if (is.null(limits) && !is.null(names(values))) {
+    force(values)
```
Comment (yutannihilation) [inline R/scale-manual.r:136]: Why is this line needed? Wasn't `values` forced in the last `if` section?

**Review by yutannihilation — COMMENTED** [review]
Thanks, while I'm not fully confident yet, this pull request looks the right direction to me. Manual scales are somewhat special, so it's probably a good idea to treat them specially, rather than modifying the whole `Scale`.

File: R/scale-manual.r (hunk 1 above)
Comment (teunbrand) [reply R/scale-manual.r]: You're right, it must have slipped my attention. Apologies!

Comment (yutannihilation) [reply R/scale-manual.r]: I see, thanks.

**Comment by clauswilke** [comment]
I have to admit I'm quite confused by this. It's hard to keep track of the many different issues that were filed. I thought we had merged the `limits = force` solution. Is that not the case?

In any case, @teunbrand, could you add tests and a news item, just like in #4547? It would be good to have a test for every single problem that has shown up.

And also, #4471 was exclusive to manual scales, so a fix exclusive to manual scales would be preferred. This argues in favor of this PR over #4547.

**Comment by yutannihilation** [comment]
> It's hard to keep track of the many different issues that were filed.

I too feel so...

> I thought we had merged the `limits = force` solution.

If I understand correctly, this pull request is based on the same idea. This comment (https://github.com/tidyverse/ggplot2/issues/4511#issuecomment-888392422) might help:

> In addition: perhaps, if the aim of #4471 was to allow fewer elements, then the line above could be changed to:

**Comment by yutannihilation** [comment]
> In any case, @teunbrand, could you add tests and a news item, just like in #4547? It would be good to have a test for every single problem that has shown up.

I hope we can narrow #4547 to only the tests and merge it. Even if we'll choose this pull request, it's good if we can count these prior works on #4547 as a contribution.

**Comment by clauswilke** [comment]
Ah, Ok. I have no concerns about this, except the order of merging. Can we merge the test before merging the fix?

**Comment by yutannihilation** [comment]
Yes, let's merge the tests first.

**Comment by banfai** [comment]
@clauswilke @yutannihilation 
I agree, this is a better solution. I have modified #4547 as requested to contain the tests only.

**Comment by banfai** [comment]
@teunbrand
This PR should include an update to NEWS.md
It could be changed to something like this, if you agree to share credit 😉 
```
* Fixes issue with `scale_*_manual` introduced in #4471 where extra elements
  in `values` or their ordering changes the legend
  (@teunbrand, @banfai, #4511, #4534, #4556, #4545, #4616)
```

**Comment by teunbrand** [comment]
@banfai Yes of course, I'm sorry for the conflict with your PR and the hassle that came with it!

**Comment by yutannihilation** [comment]
@banfai 
Ah, thanks, I was going to ask to include you in NEWS otherwise :)

@teunbrand
Now that #4547 is merged, could you merge the current master for tests?

File: NEWS.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 # ggplot2 (development version)
 
+* `scale_*_manual()` no longer displays extra legend keys, or changes their 
+  order, when a named `values` argument has more items than the data. The 
+  previous behaviour can be replicated by using 
+  `scale_*_manual(values = vals, limits = names(vals))`. (@teunbrand, @banfai, 
+  #4511, #4534, #4556, #4545, #4616)
```
Comment (yutannihilation) [inline NEWS.md:7]: While there are many issues filed, I think they can be summarised to these two ones.

```suggestion
  #4511, #4534)
```

File: R/scale-manual.r [hunk 3]
```diff
@@ -132,8 +132,8 @@ manual_scale <- function(aesthetic, values = NULL, breaks = waiver(), ..., limit
     force(values)
   }
 
-  if (is.null(limits)) {
-    limits <- names(values)
+  if (is.null(limits) && !is.null(names(values))) {
+    limits <- function(x) {intersect(x, names(values))}
```
Comment (yutannihilation) [inline R/scale-manual.r:136]:
```suggestion
    limits <- function(x) intersect(x, names(values))
```

Tidyverse style guide says:

> `{` should be the last character on the line. 
(https://style.tidyverse.org/syntax.html#indenting)

So, it's probably recommended to write this as either

``` r
    limits <- function(x) intersect(x, names(values))
```

or

``` r
    limits <- function(x) {
      intersect(x, names(values))
    }
```

**Review by yutannihilation — COMMENTED** [review]
Minor comments. Let me for a while to check if this fixes all the problems that are reported so far.

I'd also like to request to add a short comment on the `if` branch why `limits` needs to be a function.

**Comment by yutannihilation** [comment]
Hmm, I have no idea what this error is...

```
══ Failed tests ════════════════════════════════════════════════════════════════
── Error (test-scale-manual.r:119:3): limits and breaks (#4619) ────────────────
Error: Error: argument is missing, with no default
Error: Error: R CMD check found ERRORs
Execution halted
Backtrace:
    █
 1. └─ggplot2::scale_colour_manual(...) test-scale-manual.r:119:2
 2.   └─ggplot2:::manual_scale(aesthetics, values, breaks, ..., na.value = na.value)
 3.     └─ggplot2::discrete_scale(...)
 4.       └─ggplot2::ggproto(...)

[ FAIL 1 | WARN 0 | SKIP 66 | PASS 1619 ]
```

**Comment by yutannihilation** [comment]
Ah, I should have caught that trailing comma, sorry.

**Comment by teunbrand** [comment]
Alright this seems all in its intended state for now :) Thanks for the feedback @yutannihilation!

>  Let me for a while to check if this fixes all the problems that are reported so far.

That seems like a good plan to me, please let me know if you find something.

**Comment by banfai** [comment]
@teunbrand This was a better solution, I'm glad we could reconcile the two PRs.
@yutannihilation Sorry about the trailing comma error, there used to be an `na.values` arg there and it didn't throw an error on my machine.

**Comment by yutannihilation** [comment]
@banfai 
No worries, it seems something changed between R 3.4 and 3.5. It's what only the CI can find, so it's not your fault. I was a bit too lazy to check the details of the failures.

File: NEWS.md [hunk 4]
```diff
@@ -1,5 +1,11 @@
 # ggplot2 (development version)
 
+* `scale_*_manual()` no longer displays extra legend keys, or changes their 
+  order, when a named `values` argument has more items than the data. The 
+  previous behaviour can be replicated by using 
+  `scale_*_manual(values = vals, limits = names(vals))`. (@teunbrand, @banfai, 
+  #4511, #4534)
```
Comment (yutannihilation) [inline NEWS.md:7]: The word "previous behaviour" might be a bit confusing to those who want further previous behaviour, i.e. behaviour before v3.3.4. How about describing it a bit more explicitly something like "To display all the `values` on the legend, use `scale_*_manual(values = vals, limits = names(vals))`"? What do you think?

**Review by yutannihilation — APPROVED** [review]
I confirmed this solves all the related issues. Now this looks good to me!

c.f. just in case this is useful for someone, this is my quick summary of the issues.
https://gist.github.com/yutannihilation/3d0954fe2a183c08288807215e3c9705

File: NEWS.md (hunk 4 above)
Comment (clauswilke) [reply NEWS.md:7]: Agreed, "previous behavior" is confusing. Better to describe the specific behavior. I like the proposed sentence.

**Comment by yutannihilation** [comment]
Perfect, thank you so much!
